doc/design/docker-cp.md

```diff
@@ -0,0 +1,146 @@
+
+
+```
+/container/filesystem?ID=
+```
+
+For a copy *to* the filesystem of a container we will do a `PUT` operation
+
+```
+/contrainer/filesystem?ID= PUT
+
+Request Model : 
+
+* path - container filesystem target for the copy operation
+
+* ID - this is the target container
+
+* overwriteDir - flag for whether or not we are directed to overwrite existing directories 
+
+Response Model :
+
+* Error - error response from the portlayer
+
+```
+
+For a copy *from* the filesystem of a container we will do a `GET` operation. 
+
+```
+/container/filesystem?ID=
+
+
+Request Model : 
+
+* ID - this is the target container
+
+* path - target path for the copy operation
+
+Response Model : 
+
+* Error - error response from the portlayer
+
+```
+
+### Personality design
+
+The personality has two endpoints that are called depending on the to/from situation above. The portlayer will divide the behavior of to/from as two separate endpoints. These endpoints will behave in an `Export/Import` behavior. There will be several checks during this process. 
+
+Error codes expected by docker : 
+
+```
+200 
+
+400
+
+403 - in the event of a readonly filesystem
+
+404
+
+409
+
+500
+
+```
+
+Docker does not do much in the way of string checking as far as the path is concerned. This is likely not a big concern for the personality as the actual copy commmand will fail if a bad path string is provided. 
+
+The personality code should be fairly straight forward, the goal is to call the portlayer for the copy while passing the tar stream either to the portlayer or from it. The best way to do this is to hook the writiers together using a PipeReader and then shepherding the data in the direction it needs to go. 
+
+## Portlayer Design
+
+ There will be two situations to be concerned with in the portlayer. When the container is *on* and when the container is *off*.
+ 
+ 
+### Off Container Scenario
+ 
+ This is the case where guest tools is not needed since the guest in off. This means that all action will need to occur on the VCH. There are several concerns and many of them are listed in this ticket [here](https://github.com/vmware/vic/issues/5295)
+ 
+ The most concern surrounding this piece of work will be surrounding VCH restart and confirming that the disk is still left in a good state if a copy was occuring mid restart. Additionally we need to block a container from starting when we have the r/w layer mounted. While that container is blocked from being started we might look at the possibility of returning a 500 error for a start invocation on a blocked container. We must also block other copy operations until the current one is successful, one major reason is because we cannot do any kind of sanity checking on available space with concurrent writes to the same disk. Blocking operations until another is completed is ideal, since we work in a highly automated environment.
+ 
```

Comment:
I would remove "The merrits for calling `fsck` have not been determined yet.  Investigation will be needed to see what kind of writes and operations could make us susceptible to needing `fsck`" and simply say,

We may need to further investigate whether we will also need `fsck` to check filesystem consistency.